### DISCUSSION TITLE
General Questions about the JS SDK

### DISCUSSION THREAD

**Discussion by BenKurrek** [near/near-sdk-js#32] [Q&A] [answered]
I’ve got a couple question regarding the JS SDK. I’ve been trying to get a contract running for so long and I’m reaching a point where I think i’m really stuck. I put a TLDR at the bottom of this message before my contract since I know it’s a bit long.

I’ve got a very simple contract (I think) and I can’t seem to get anything to work if I’m passing in values into a constructor. I’ve attached my contract at the bottom. 

When deploying, I’ve tried almost every single combination of trying to pass in an initial total supply of 10. What seems the most correct (since it mirrors the CLI) is this command to call the initialization function:
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ init '{"totalSupply": 10}') --base64 --amount 1
```
But this throws an interesting error: `ExecutionError: 'Smart contract panicked: value is not iterable\n'`  I’ve tried doing some research but couldn’t really find anything helpful. The only way I could figure out how to not make this throw is by passing in empty arguments. (I’m also a little confused as to what the super() function is used for and why it’s needed.)
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ init '') --base64 --amount 1
```
which returned:
```
Scheduling a call: jsvm.testnet.call_js_contract(anMtNC5leGFtcGxlcy5iZW5qaW1hbi50ZXN0bmV0AGluaXQA) with attached 1 NEAR
Doing account.functionCall()
Receipts: 4c1izr7ruGUYQBybyHudEr5bBE5W9W5eNdpHGWpg9PDp, 4yX5hYiLTAfSQ5PhaMnM4kww371NFkAurGx2fCkd4MFN
	Log [jsvm.testnet]: Constructor called!
	Log [jsvm.testnet]: Super Called!
	Log [jsvm.testnet]: Total supply set
Transaction Id 4w19tzchV7WS2wRQomxeGFy9ZMmqoDFPXAcG8vAAxJVF
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.testnet.near.org/transactions/4w19tzchV7WS2wRQomxeGFy9ZMmqoDFPXAcG8vAAxJVF
```
This however sets the total supply to null and breaks the contract. The workaround I’m using for this is to get rid of the parameter being passed into the constructor and just hard set `this.totalSupply = 0`  which seems to do the trick. After deploying and testing the initialization function, everything seemed to work just fine and when I called the `getTotal()` function, it correctly returned 0. 

What was confusing, however, was when I called the `getTotal() function`, the env.logs I placed in the constructor got fired again:
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ getTotal '') --base64
```  
returns:
```
Scheduling a call: jsvm.testnet.call_js_contract(anMtNC5leGFtcGxlcy5iZW5qaW1hbi50ZXN0bmV0AGdldFRvdGFsAA==)
Doing account.functionCall()
Receipt: D4NrnwxCv5gzhpDex4WyCUkn4tpjV9HwGyb4vAEo5fpg
	Log [jsvm.testnet]: Constructor called!
	Log [jsvm.testnet]: Super Called!
	Log [jsvm.testnet]: Total supply set
	Log [jsvm.testnet]: Get total called!
Transaction Id ApsbCUTeen3tNcpM392LRa8vXePMioRWa9gdVr3XeTij
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.testnet.near.org/transactions/ApsbCUTeen3tNcpM392LRa8vXePMioRWa9gdVr3XeTij
```
I then wondered if it was resetting the totalSupply everytime I ran the getTotal function. I tested this by attempting to increment the totalSupply and then calling the getTotal function again. I ran into the same problem of not being able to pass anything into the increment function but instead of receiving the value is not iterable error, it simply thinks I passed in null:
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ increaseTotal '{"n": 10}') --base64 --amount 1
```
which returns:
```
Scheduling a call: jsvm.testnet.call_js_contract(anMtNC5leGFtcGxlcy5iZW5qaW1hbi50ZXN0bmV0AGluY3JlYXNlVG90YWwAeyJuIjogMTB9) with attached 1 NEAR
Doing account.functionCall()
Receipts: 5rAD4K7E5rD5ztYk72pYYbPDjNBhVYsQZ9GrU4b91cDX, 8J9XTEJgMC2XArERBrsHZQEbE5TL2CqjU4rKVS4mpQzX
	Log [jsvm.testnet]: Constructor called!
	Log [jsvm.testnet]: Super Called!
	Log [jsvm.testnet]: Total supply set
	Log [jsvm.testnet]: Increase total called with:
	Log [jsvm.testnet]: NaN
Transaction Id FAXVXdfeKhGktfpCHw9rWqG7imMGE86cRcagFCwytsx5
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.testnet.near.org/transactions/FAXVXdfeKhGktfpCHw9rWqG7imMGE86cRcagFCwytsx5
```
After checking the current totalSupply using getTotal, it returned null:
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ getTotal '') --base64 
```
returns:
```
Scheduling a call: jsvm.testnet.call_js_contract(anMtNC5leGFtcGxlcy5iZW5qaW1hbi50ZXN0bmV0AGdldFRvdGFsAA==)
Doing account.functionCall()
Receipt: 6B8FLRcyM4CuzSpKy6DtVxKnXYgBFRYBQxmqfdQy3cbR
	Log [jsvm.testnet]: Constructor called!
	Log [jsvm.testnet]: Super Called!
	Log [jsvm.testnet]: Total supply set
	Log [jsvm.testnet]: Get total called!
Transaction Id CgVvDi8if2Vz4vphdiuu3EqhygNLtAtsSC71qBMWq7S4
To see the transaction in the transaction explorer, please open this url in your browser
https://explorer.testnet.near.org/transactions/CgVvDi8if2Vz4vphdiuu3EqhygNLtAtsSC71qBMWq7S4
null
```

TLDR:
- how do I pass in values into a function?
- what is the super() function and why is it needed in the constructor?
- how come my getTotal()function runs through the constructor everytime it’s called?
- how do I invoke a view method without signing a transaction?
- I’ll also cross-post one of my previous questions here so it’s all in one place:
- How is state cleanup handled? If I wanna redeploy a contract to an account, I can’t simply delete the account and recreate it - because the state is stored on the jsvm contract correct?

```bash
import {NearContract, NearBindgen, call, view} from '../../sdk'
// import { init } from './build/project.js';

@NearBindgen
class MyContract extends NearContract {
    constructor(totalSupply) {
        env.log("Constructor called!")
        super()
        env.log("Super Called!")
        this.totalSupply = totalSupply
        env.log("Total supply set")
    }

    @call
    increaseTotal(n) {
        env.log("Increase total called with: ", n)
        this.totalSupply += n
        env.log(this.totalSupply)
        return this.totalSupply
    }

    @view
    getTotal() {
        env.log("Get total called!")
        return this.totalSupply
    }
}
```

**Comment by BenKurrek** [comment]
From @bucanero: 

about the super(), as many OOP languages do, it’s used to call the parent class’ constructor.

Example:
If you have a parent class Tool , and a class Hammer that inherits from Tool, if the Hammer() constructor calls super(), the Tool() constructor gets executed.

That’s the general theory, but I guess you need it in your smart contract as some initialization from the NearContract class is required.

**Comment by BenKurrek** [comment, marked as answer]
From @ailisp: 

In response to "how do I pass in values into a function? and also the error about passing value to constructor":  
You need to pass argument as array:
```bash
near call $JSVM_ACCOUNT call_js_contract --accountId $JSPROJ --args $(node encode_call.js $JSPROJ init '[10]') --base64 --amount 1
```
If there were more than one arguments, for example: (totalSupply, name) , pass it as: [10, "alice"]

In response to "how come my getTotal()function runs through the constructor everytime it’s called?": 

Your observation is right. This is an intentional designed code generation to adapt JS's OOP model to NEAR's contract function call lifecycle. And it's the same design as near-sdk-rs's near_bindgen macro.

In JS, there're classes and objects. In a smart contract, the entry point can only be a function. When you call `getTotal` with `near call ... encode_call.js ...`, you are not calling the `Counter.getTotal` method, because there's no Counter yet. Instead, it's calling a generated `getTotal` function, which does the following:
1. create a Counter object (this is where constructor get called)
2. reset Counter's properties by reading the state from chain, so the `Counter.count` is the last value saved by `increaseTotal/decreaseTotal` on chain, instead of the initial value
3. call Counter.getTotal

In response to "how do I invoke a view method without signing a transaction?": 

Call view_js_contract method. This is currently not possible with near-cli due to this: https://github.com/near/near-cli/issues/953, but it's doable with near-api-js, see examples/project/near-workspaces/__tests__/project.ava.ts for an example

In response to "How is state cleanup handled? If I wanna redeploy a contract to an account, I can’t simply delete the account and recreate it because the state is stored on the jsvm contract correct?": 

You will need to write a js contract, do a couple calls to jsvm_storage_delete to clear state

**Comment by volovyks** [comment]
@BenKurrek this discussion looks mostly stale. We have new `near js <command>` and new APIs in `near-sdk-js`. Should we close it?
